astropy/io/misc/asdf/tags/coords/coords.py

```diff
@@ -0,0 +1,57 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+# -*- coding: utf-8 -*-
+
+from asdf.yamlutil import custom_tree_to_tagged_tree
+
+from astropy.units import Quantity
+from astropy.coordinates import ICRS, Longitude, Latitude, Angle
+from astropy.tests.helper import assert_quantity_allclose
+
+from ...types import AstropyType
+from ..unit.quantity import QuantityType
+
+
+__all__ = ['ICRSCoordType']
+
+
+class ICRSCoordType(AstropyType):
+    name = "coords/icrs_coord"
+    types = ['astropy.coordinates.ICRS']
+    requires = ['astropy']
+    version = "1.0.0"
+
+    @classmethod
+    def from_tree(cls, node, ctx):
+        angle = Angle(QuantityType.from_tree(node['ra']['wrap_angle'], ctx))
+        wrap_angle = Angle(angle)
+        ra = Longitude(
```

Comment:
Sorry to be a nuisance, but could you do this `Longitude` and the `Latitude` just below as well? I.e.,
```
ra = Longitude(node['ra'], wrap_angle=wrap_angle)
dec = Latitude(node['dec'])
```
After that, I think it's done, so I guess you'll need to do the little dance of the merges at the two side.

Reply:
When I try to do that I get an error. The values in `node['ra']` are just floats, so I'm not sure if it will work that way, although maybe I'm missing something.



[test_failure.txt](https://github.com/astropy/astropy/files/1528311/test_failure.txt)

Reply:
Ah, now I see that here you are accessing two keys. Would it then make more sense to use the same abstraction as for `wrap_angle` and do `Longitude(QuantityType.from_tree(node['ra']), wrap_angle=wrap_angle)`? That would seem to keep the implementation more in one place? (and take automatic care of the case that the node is already a `Quantity`). As a by-product, it would seem to make the code more future-proof, since `QuantityType` does handle arrays as well.

(In fact, maybe go even further and create `AngleType`, `LongitudeType`, and `LatitudeType`, which subclass from `QuantityType` - though it does seem one is repeating a lot of stuff from the actual initializers...)

Reply:
That's a good idea. Actually one of the reasons for moving these things to Astropy is that hopefully we can provide explicit support for more types, as you suggest.

Reply:
Actually after looking into it a little more closely, I realized that supporting the implementation you suggested is going to require a schema change. I'm not opposed to this since it would probably help make this a little cleaner. However, that's not the goal of this PR and I would rather submit it as a separate change if you don't mind.

Reply:
I agree with @drdavella . This is just the first step and is the only schema currently available in asdf. I'd prefer to merge this so that we know the mechanism works. And then come back and write/enhance the schemas for the other coordinates objects. Should probably file issues for this.

Reply:
On a related note, I'd eventually like to explore how much we can leverage ECSV machinery for ASDF tag implementations in Astropy, but that is also work for another day (cc @taldcroft).

Reply:
Maybe just do the `QuantityType` then, not the subclassing, i.e., do it the same way within this PR.  I'm totally fine with postponing new `AngleType`, etc., to another PR (indeed, one should first make sure it even makes sense).

Reply:
@mhvk adding the `QuantityType` here as you proposed is the part that will require a schema change. There were reasons for doing it the original way, but I agree that in the long term it would be better to use `QuantityType`. For now, hopefully it's okay to merge without the schema update.